wip (#4293)
facebook/react-native-website: website/blog/2024-10-23-The-New-Architecture-is-Here.md

### Gradual Migration

The New Architecture in React Native is a breaking change, but the migration process has been made easier with tools and [interop layers](https://github.com/reactwg/react-native-new-architecture/discussions/135).

The concurrent features of React are [only enabled in parts of the application that use them](https://react.dev/blog/2022/03/29/react-v18#gradually-adopting-concurrent-features), and modules and components can be migrated to be "New Architecture compatible" without requiring modification. However, there are [known limitations](https://github.com/reactwg/react-native-new-architecture/discussions/175) to the interop layers, and some modules and components may need to be [properly migrated](TODO).

The interop layers will be phased out in future releases of React Native. Meanwhile, if you find a bug with them, please [open them using this template](https://github.com/facebook/react-native/issues/new?assignees=&labels=Needs%3A+Triage+%3Amag%3A%2CType%3A+New+Architecture&projects=&template=new_architecture_bug_report.yml).

## New Features

> The New Architecture includes full support for React 18, concurrent features, and useLayoutEffect in React Native. For a full list of React 18 features, please see the [React 18 blog post](https://react.dev/blog/2021/12/17/react-conf-2021-recap#react-18-and-concurrent-features).

Reviewer: ```suggestion
The New Architecture includes full support for React 18, concurrent features, and `useLayoutEffect` in React Native. For a full list of React 18 features, please see the [React 18 blog post](https://react.dev/blog/2021/12/17/react-conf-2021-recap#react-18-and-concurrent-features).
```

The RFC follows the specifications described in the [HTML specifications](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model) and it describes how React Native should perform tasks on the JavaScript thread.

Implementing a well-defined event loop helps closing the gaps between web development and native development: the behavior of a React Native application is now closer to the behavior of a React application, allowing you to really learn once, and write anywhere.

The event loop brings many benefits to React Native. Among those, we have:

- A predictable behavior of the framework, that leads to more predictable and performant apps
- A better alignment of React native with the Web
- It enables the implementation of more compliant web specifications
- It lays the foundation for new features like Microtasks, `useLayoutEffect` and others features that will be implemented in the future.

Reading layout in useLayoutEffect

> The event loop allowed us to implement the useLayoutEffect hooks in React Native. This is the perfect place to retrieve the most accurate measurements of the views in your UI.

Reviewer: ```suggestion
The event loop allowed us to implement the `useLayoutEffect` hooks in React Native. This is the perfect place to retrieve the most accurate measurements of the views in your UI.
```

In the legacy architecture, due to its asynchronous nature, we could not have synchronous layouts. To retrieve the latest layout information of a view, we had to wait before the view was actually rendered. With the New Architecture we can now use useLayoutEffect to implement synchronous layouts.

Let’s have a look at this example: rendering a tooltip next to a view that moves on the screen.

> With the legacy Architecture, on the left, we had to wait for the view to be completely rendered before we could fetch its layout information and render the tooltip next to it. This leads to a user experience where the tooltip would lag behind the view.

Reviewer: ```suggestion
With the legacy architecture, on the left, we had to wait for the view to be completely rendered before we could fetch its layout information and render the tooltip next to it. This leads to a user experience where the tooltip would lag behind the view.
```

<div className="TwoColumns TwoFigures">
<figure>
 <img src="/img/new-architecture/async-on-layout.gif" alt="A view that is moving to the corners of the viewport and center with a tooltip rendered either above or below it. The tooltip is rendered after a short delay after the view moves" />
 <figcaption>Asynchronous measurement and render of the ToolTip. [See code](https://gist.github.com/lunaleaps/eabd653d9864082ac1d3772dac217ab9).</figcaption>
</figure>
<figure>
 <img src="/img/new-architecture/sync-use-layout-effect.gif" alt="A view that is moving to the corners of the viewport and center with a tooltip rendered either above or below it. The view and tooltip move in unison." />
 <figcaption>Synchronous measurement and render of the ToolTip. [See code](https://gist.github.com/lunaleaps/148756563999c83220887757f2e549a3).</figcaption>
</figure>
</div>

### useLayoutEffect

Building on the Event Loop and the ability to read layout synchronously, in the New Architecture we added proper support for `useLayoutEffect` in React Native.

> In the old archiectur, you needed to use the asyncronous `onLayout` event to read layout information of a view (which was also asynchronous). As a result there would be at least one frame where the layout was incorrect until the layout was read and updated, causing issues like tooltip placed in the wrong position:

Reviewer: ```suggestion
In the old architecture, you needed to use the asynchronous `onLayout` event to read layout information of a view (which was also asynchronous). As a result there would be at least one frame where the layout was incorrect until the layout was read and updated, causing issues like tooltip placed in the wrong position:
```

I see we're sometimes referring it as "old architecture", and sometimes as "legacy" -- should we conform on one?

```tsx
// ❌ async onLayout after commit
const onLayout = React.useCallback(event => {
  // ❌ async callback to read layout
  ref.current?.measureInWindow((x, y, width, height) => {
    setPosition({x, y, width, height});
  });
}, []);

// ...
<ViewWithTooltip
  onLayout={onLayout}
  ref={ref}
  position={position}
/>;
```

The New Architecture fixes this by allowing synchronous access to layout information in `useLayoutEffect`:

```tsx
// ✅ sync layout effect during commit
useLayoutEffect(() => {
  // ✅ sync call to read layout
  const rect = ref.current?.getBoundingClientRect();
  setPosition(rect);
}, []);

// ...
<ViewWithTooltip ref={ref} position={position} />;
```

This change allows you to read layout information synchronously and update the UI in the same frame, allowing you to position elements correctly before they are displayed to the user.

## How to Upgrade

React Native 0.76 is the first version of React Native that leverages the full power of [React 18 and its concurrent features](https://react.dev/blog/2021/12/17/react-conf-2021-recap#react-18-and-concurrent-features).

> For this reason, we expect that you need to apply some changes to your JavaScript code to properly migrate to React 18 and its semantics. For more information on how to migrate to React 18’s concurrent feature, follow the [How to Upgrade to React 18](https://react.dev/blog/2022/03/08/react-18-upgrade-guide).

Reviewer: ```suggestion
To migrate your JavaScript code to React 18 and its semantics, follow the [How to Upgrade to React 18](https://react.dev/blog/2022/03/08/react-18-upgrade-guide) guide.
```

> For what concerns the native side of your app, version 0.76 comes with some breaking changes that you need to consider when upgrading your app. The list of breaking changes are described in the [React Native 0.76 release blogpost post](/blog/2024/10/23/release-0.76-new-architecture#breaking-changes-1) and in the [React Native changelog](https://github.com/facebook/react-native/blob/main/CHANGELOG.md).

Reviewer: ```suggestion
For native migration, 0.76 comes with some breaking changes that you need to consider when upgrading your app. The list of breaking changes are described in the [React Native 0.76 release blogpost post](/blog/2024/10/23/release-0.76-new-architecture#breaking-changes-1) and in the [React Native changelog](https://github.com/facebook/react-native/blob/main/CHANGELOG.md).
```

> Aside from the breaking changes described above, the migration to the New Architecture is transparent.

Reviewer: ```suggestion
Aside from the breaking changes described above, the migration to the New Architecture is transparent.
```
This sentence sounds a bit odd and un-necessary? Could we cut?

> To migrate to the New Architecture, follow the usual steps as they are illustrated in the [upgrade helper](https://react-native-community.github.io/upgrade-helper/). Once you carry out these steps, your application will run on the New Architecture.

Reviewer: How is this different from upgrading to 0.76 since it should be default enabled? Or am I missing something?

Author: I rewrote this whole section, what do you think?

### Libraries

> We verified that all the most used libraries are compatible with the New Architecture. Specifically, we monitored all the libraries with more 200K weekly downloads (that's \~10% of React Native weekly downloads) and we verified that they work well with the New Architecture

Reviewer: ```suggestion
We verified that all the most used libraries are compatible with the New Architecture. Specifically, we monitored all the libraries with more than 200K weekly downloads (that's \~10% of React Native weekly downloads) and we verified that they work well with the New Architecture.
```

To opt-out from the New Architecture:

- On Android, modify the `android/gradle.properties` file and turn off the `newArchEnabled` flag

```diff title=”gradle.properties”
-newArchEnabled=true
+newArchEnabled=false
```

- On iOS, you can reinstall the dependencies running the command:

```shell
RCT_NEW_ARCH_ENABLED=0 bundle exec pod install
```

## Thanks

Delivering the New Architecture to the OSS community has been a huge effort which took us several years of research and development. We want to take a moment to thank all the current and past members of the React team that helped us achieve this result.

We are also extremely grateful to all the partners that collaborated with us to make this happen. Specifically, we would like to call out:

- [Expo](https://expo.dev/), for adopting the New Architecture early on, for supporting the work on migrating the most popular libraries.
- [Software Mansion](https://swmansion.com/), for maintaining crucial libraries in the ecosystem, for migrating them to the New Architecture early and for all the help in investigating and fixing various issues.

> - [Callstak](https://www.callstack.com/), for maintaining crucial libraries in the ecosystem, for migrating them to the New Architecture early and for the support with the work on the Community CLI.

Reviewer 2: Callstack

---
title: 'The New Architecture is here'
authors: [reactteam]
tags: [announcement]
date: 2024-10-23
---

React Native 0.76 with the New Architecture by default is now available on npm!

> In the [0.76 release blog post](/blog/2024/10/23/release-0.76-new-architecture) we shared the list of all the significant changes included in this version. In this post, we want to give an overview of the New Architecture and how it is going to shape the future of React Native.

Reviewer 3: ```suggestion
In the [0.76 release blog post](/blog/2024/10/23/release-0.76-new-architecture), we shared a list of significant changes included in this version. In this post, we provide an overview of the New Architecture and how it shapes the future of React Native.
```

> The New Architecture adds support for modern React features including full support for [Suspense](https://react.dev/blog/2022/03/29/react-v18#new-suspense-features), [Transitions](https://react.dev/blog/2022/03/29/react-v18#new-feature-transitions), [automatic batching](https://react.dev/blog/2022/03/29/react-v18#new-feature-automatic-batching), and [`useLayoutEffect`](https://react.dev/reference/react/useLayoutEffect). The New Architecture also includes new [native module](/docs/next/fabric-native-components-introduction) and [native component](/docs/next/fabric-native-components-introduction) systems that enable writing type-safe modules with direct access to native without a bridge.

Reviewer 3: ```suggestion
The New Architecture adds full support for modern React features including [Suspense](https://react.dev/blog/2022/03/29/react-v18#new-suspense-features), [Transitions](https://react.dev/blog/2022/03/29/react-v18#new-feature-transitions), [automatic batching](https://react.dev/blog/2022/03/29/react-v18#new-feature-automatic-batching), and [`useLayoutEffect`](https://react.dev/reference/react/useLayoutEffect). The New Architecture also includes new [native module](/docs/next/fabric-native-components-introduction) and [native component](/docs/next/fabric-native-components-introduction) systems that let you write type-safe code with direct access to native interfaces without a bridge.
```

> The release is possible thanks to a ground up rewrite of React Native we’ve been working on since 2018, and we’ve taken special care to make the New Architecture a gradual migration for most apps. In 2021, we created [the New Architecture Working group](https://github.com/reactwg/react-native-new-architecture/) to work with the community to support the New Architecture and ensure a smooth upgrade experience for the entire React ecosystem.

Reviewer 3: ```suggestion
This release is the result of a ground up rewrite of React Native we’ve been working on since 2018, and we’ve taken special care to make the New Architecture a gradual migration for most apps. In 2021, we created [the New Architecture Working Group](https://github.com/reactwg/react-native-new-architecture/) to collaborate with the community on ensuring a smooth upgrade experience for the entire React ecosystem.
```

> We expect most apps can adopt the New Architecture with the same effort as any other release. The most popular React Native libraries already include support for the New Architecture, and for libraries that do not support the New Architecture, we developed an automatic interop layer with the old architecture.

Reviewer 3: ```suggestion
Most apps will be able to adopt React Native 0.76 with the same level of effort as any other release. The most popular React Native libraries already support the New Architecture. The New Architecture also includes an automatic interoperability layer to enable backward compatibility with libraries targeting the old architecture.
```

<!--truncate-->

> In case you missed it, we shared a lot of this vision over the past years:

Reviewer 3: ```suggestion
Over the past several years of development, our team has publicly shared our vision for the New Architecture. If you missed any of these presentations, check them out here:
```

- [React Native EU 2019 \- The New React Native](https://www.youtube.com/watch?v=52El0EUI6D0)
- [React Conf 2021 \- React 18 Keynote](https://www.youtube.com/watch?v=FZ0cG47msEk)
- [App.js 2022 \- Bringing the New React Native Architecture to the OSS Community](https://www.youtube.com/watch?v=GJxL2aiIX3Q)
- [React Conf 2024 \- Day 2 keynote](https://www.youtube.com/watch?v=Q5SMmKb7qVI)

## What is the New Architecture

> The most important change in this release is something most users should not need to think about: how React Native works. We have completely rewritten the major systems of React Native including how components are rendered, how JavaScript communicates with native, and how work is scheduled across different threads. We call all of these architecture improvements the New Architecture.

Reviewer 3: This paragraph feels to me like it is trying too hard to build up the New Architecture, even though we've already introduced it above and in the heading for this section.

I would just go right into it.

```suggestion
The New Architecture is a complete rewrite of the major systems that underpin React Native, including how components are rendered, how JavaScript abstractions communicates with native abstractions, and how work is scheduled across different threads. Although most users should not have to think about how these systems work, these changes bring improvements and new capabilities.
```

Author: Love it

> In the old architecture, React Native communicated with native with an asynchronous bridge. To render a component or call a native function, React Native needed to push native functions calls to a queue in the bridge, and process them asynchronously. This architecture had many benefits. Since all work was done on a background thread, the main thread was never blocked for rendering updates or handling native module function calls.

Reviewer 3: ```suggestion
In the old architecture, React Native communicated with the native platform using an asynchronous bridge. To render a component or call a native function, React Native needed to serialize and enqueue native functions calls with the bridge, which would be processed asynchronously. This architecture had many benefits. Since all work was done on a background thread, the main thread was never blocked for rendering updates or handling native module function calls.
```

### Full support for Suspense

Suspense lets you declaratively specify the loading state for a part of the component tree if it’s not yet ready to be displayed:

```
<Suspense fallback={<Spinner />}>
  <Comments />
</Suspense>
```

We introduced a limited version of Suspense several years ago, and React 18 added full support. However, until now React Native was not able to support concurrent rendering for Suspense.

The New Architecture includes full support for Suspense introduced in React 18. This means that you can now use Suspense in React Native to handle loading states for your components, and the suspended content will render in the background while the loading state is displayed, giving higher priority to user input on visible content.

### Event Loop

The New Architecture allowed us to implement a well-defined event loop processing model, as described in this [RFC](https://github.com/react-native-community/discussions-and-proposals/blob/main/proposals/0744-well-defined-event-loop.md).

> The RFC follows the specifications described in the [HTML specifications](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model) and it describes how React Native should perform tasks on the JavaScript thread.

Reviewer 3: ```suggestion
The RFC follows the specifications described in the [HTML Standard](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model) and it describes how React Native should perform tasks on the JavaScript thread.
```

## What is the New Architecture

The most important change in this release is something most users should not need to think about: how React Native works. We have completely rewritten the major systems of React Native including how components are rendered, how JavaScript communicates with native, and how work is scheduled across different threads. We call all of these architecture improvements the New Architecture.

In the old architecture, React Native communicated with native with an asynchronous bridge. To render a component or call a native function, React Native needed to push native functions calls to a queue in the bridge, and process them asynchronously. This architecture had many benefits. Since all work was done on a background thread, the main thread was never blocked for rendering updates or handling Native Module function calls.

The approach had major limitations. Some work needs to respond synchronously to user input, potentially interrupting any in progress work, in order to provide the best experience possible to users. Since the old architecture was only asynchronous, we needed to rewrite it to allow for both asynchronous and synchronous updates.

> 

Reviewer 2: Synchronous user input is not (yet) supported in the new architecture, so we don't have synchronous update. What we do have now is the ability for user input to interrupt other long-running tasks.

Author: It's nuanced, but this isn't talking specifically about dispatching events synchronously. What it does in the New Architecture today is that read renders discrete events synchronously (without yielding) interrupting lower priority updates like transitions. But I can tweak this to talk about rendering synchronously instead of events.

Author: Fixed in the other branch.
